src/Boson_Sampling_Utilities.py

```diff
@@ -121,3 +122,108 @@ def generate_lossy_inputs(initial_state: ndarray, number_of_particles_left: int)
             lossy_inputs_list.append(lossy_input)
 
     return lossy_inputs_list
+
+
+def calculate_number_of_possible_n_particle_m_mode_output_states(n: int, m: int) -> int:
+    """
+        Calculates the number of possible output states with n particles placed around m modes. This is basically
+        the same answer as to in how many possible combinations can we put n objects in m bins. It's also a dimension
+        of n-particle m-mode bosonic space. Stars-and-bars argument applies here.
+    """
+    return binom(n + m - 1, n)
+
+
+class ChinHuhPermanentCalculator:
+
+    def __init__(self, matrix: ndarray, input_state: List[int] = [], output_state: List[int] = []):
+        self.__matrix = matrix
+        self.__input_state = input_state
+        self.__output_state = output_state
+
+    @property
+    def matrix(self) -> ndarray:
+        return self.__matrix
+
+    @matrix.setter
+    def matrix(self, matrix) -> None:
+        self.__matrix = matrix
+
+    @property
+    def input_state(self) -> List[int]:
+        return self.__input_state
+
+    @input_state.setter
+    def input_state(self, input_state) -> None:
+        self.__input_state = input_state
+
+    @property
+    def output_state(self) -> List[int]:
+        return self.__output_state
+
+    @output_state.setter
+    def output_state(self, output_state) -> None:
+        self.__output_state = output_state
+
+    def calculate_permanent_of_effective_scattering_matrix(self) -> float:
```

Comment:
Nitpick: Consider `@numba.jit(nopython=True, parallel=True)` in case you need additional performance. It looks like it could benefit nicely from it.

Reply:
Performance issues I'll be considering later -- first I had to get the methods to work properly. Thanks for the tip though!